### DISCUSSION TITLE
How to use URL path for language detection?

### DISCUSSION THREAD

**Discussion by TimoWestland** [sergiodxa/remix-i18next#135] [Q&A] [answered]
As far as I can see, there is no support for using the locale in the url (e.g. `/en/path` or `/nl/path`).

I would like my app to respect the locale in the URL if the user enters the website trough an URL that has the locale in it.

However in `server.entry.ts` doing `const lng = await i18next.getLocale(request)` will always return the fallbackLng currently.

I'm using the basic setup as described in the readme of this repo.

Is there any way to solve this?

**Comment by sergiodxa** [comment, marked as answer]
You can access that from the `params` you get in your loaders and actions, or using `useParams` hook in the UI.

In the root loader:

```ts
export async function loader({ request, params }: DataFunctionArgs) {
  // use params.locale, if it's undefined (optional segments) fallback to i18next.getLocale
  let locale = params.locale ?? await i18next.getLocale(request);
  return json({ locale })
}
```

Or if you don't want to use `i18next.getLocale` you can access it direct in the root component.

```ts
export default function Component() {
  let locale = useParams().locale ?? "en" // fallback in case of optional segments
  // more code
}
```

**Comment by TimoWestland** [comment]
Hi @sergiodxa,

Thanks for your reply! That makes a lot of sense to use the remix hook to do that. 

However, it is now still getting the wrong locale in `entry.server.ts`:

```ts
  const instance = createInstance()
  const lng = await i18next.getLocale(request)
  const ns = i18next.getRouteNamespaces(remixContext)

  console.log({ lng }) // `locale: 'en' which is default locale, however we requested /nl/
```
This doesn't break anything as the root loader will set the correct locale based on the path, however it seems a little wasteful to initiate the I18nProvider with the wrong locale first.

Also, this doesn't seem to work when navigating from `/en/` to `/nl/` using Remix `Link` without using the query param.
Appending the query param or using a regular link does work.

Do you have any suggestions to solve these issues?

**Reply by sergiodxa** [reply]
In `entry.server` you will have to get the locale from the EntryContext instead of using `getLocale`, basically if you get the locale from the route params there's no need to remix-i18next, maybe only to get `getFixedT` so you can generate translated texts on the loader.

**Comment by TimoWestland** [comment]
Another issue I see (probably a remix issue not a remix-i18next issue though), is that when I request the dutch url `/nl/path`, params is returned as an empty object, which makes it fall back to the language set in entry.server.ts, which is the default language as it still isn't able to detect the language from the path there.

request to `/en/path`

in root.ts loader: `{ params: { locale: 'en' } }`

request to `/nl/path`

in root.ts loader: `{ params: {  } }`

The route for the dutch variant is defined trough `defineRoutes` in the remix.config:

```js
  routes(defineRoutes) {
    return defineRoutes((route) => {
      route(
        'nl/rtp-nl',
        'routes/$locale.rtp-nl.tsx',
        { id: 'rtp-nl-root' },
        () => {
          route('', 'routes/$locale.rtp-nl._index.tsx', {
            id: 'index-nl',
            index: true,
          })
          route(
            'afspraak-maken',
            'routes/$locale.rtp-nl.make-appointment.tsx',
            { id: 'appointment-nl' },
          )
        },
      )
    })
```

**Reply by sergiodxa** [reply]
This is because you set `nl` not as a param but as a specific part of the segment, let's me show it with normal routes
```
routes/($locale).appointments.tsx <- this has params.locale defined
routes/nl._index.tsx <- this doesn't have params.locale because there's no param on the URL
```
So in the routes where the locale is hardcoded like `nl._index` you will not receive the param and instead should hard code it too.

**Reply by TimoWestland** [reply]
Ah yeah that makes sense. Thank you!
